### PR TITLE
feat: add script to profile latency

### LINKED ISSUE #128: Profile subnet latency
Why? : We need a unified scheme to compute token/sec for subnetworks. 
How?: We cannot use set_sub_network as that adds additional overhead of weight slicing. Instead, extract sub-network and profile it using hw-gpt-bench scripts for latency profiling.

### PR DESCRIPTION
**Pull request by rheasukthanker**
#### Reference Issues/PRs
Resolves #128

#### What does this implement/fix? Explain your changes.
Implements latency profiler

#### Minimal Example / How should this PR be tested?
Example included in the latency.py file

#### Any other comments?
I decided to go with profiling litgpt models with updated config directly, mainly because the indexing, checks in whittle might introduce some additional overhead when profiling latency

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the
terms of your choice.

### CODE REVIEW

File: whittle/metrics/latency.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    config.intermediate_size = sub_network_intermediate_size
+    config.n_head = sub_network_num_heads
+    if sub_network_query_groups is not None:
+        config.n_query_groups = sub_network_query_groups
+    if sub_network_head_size is not None:
+        config.head_size = sub_network_head_size
+    return config
+
+
+if __name__ == "__main__":
```
Comment (aaronkl) [inline whittle/metrics/latency.py:79]: Could we move this into an example?

File: whittle/metrics/latency.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            # in legacy profiler, kernel info is stored in cpu events
+            if evt.is_legacy:
+                sum_self_cuda_time_total += evt.self_cuda_time_total
+        elif evt.device_type == DeviceType.CUDA:
+            # in kineto profiler, there're events with the correct device type (e.g. CUDA)
+            sum_self_cuda_time_total += evt.self_cuda_time_total
+    return sum_self_cpu_time_total, sum_self_cuda_time_total
+
+
+def profile_model_latency(model, use_cuda=False, batch_size=8, n_samples=10):
```
Comment (aaronkl) [inline whittle/metrics/latency.py:26]: Could you add types + docstring?

File: whittle/metrics/latency.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
+import torch
+import torch.profiler
+from torch.profiler import record_function
+
+
+# Define a sample language model (e.g., a simple RNN or transformer-based model)
+def get_total_cpu_gpu_runtime(prof):
+    # from torch.autograd.profiler_utils.py (_build_table)
+    # results is in micro-seconds
+    from torch.autograd import DeviceType
```
Comment (aaronkl) [inline whittle/metrics/latency.py:10]: Is there a reason to have the import here instead on top of the file?

File: whittle/metrics/latency.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+    events = prof.events()
+    sum_self_cpu_time_total = sum([event.self_cpu_time_total for event in events])
+    sum_self_cuda_time_total = 0
+    for evt in events:
+        if evt.device_type == DeviceType.CPU:
+            # in legacy profiler, kernel info is stored in cpu events
+            if evt.is_legacy:
```
Comment (aaronkl) [inline whittle/metrics/latency.py:18]: Shouldn't we have another else block here if we don't use a legacy provider?

File: whittle/metrics/latency.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    # Summing up CPU and CUDA times from the profiler using the correct methods
+    cuda_time_us, cpu_time_us = get_total_cpu_gpu_runtime(profiler)
+
+    # Convert time to milliseconds
+    total_time_ms = (cpu_time_us + cuda_time_us) / 1000
+    model = model.cpu()
+    return total_time_ms / n_samples
+
+
+def update_config(
```
Comment (aaronkl) [inline whittle/metrics/latency.py:60]: Do we need this function here? If it just belongs to the example below, I would also move it into the example folder?

File: whittle/metrics/latency.py (hunk 2 above)
Comment (aaronkl) [inline whittle/metrics/latency.py:26]: Maybe for  a future PR: We are a bit inconsistent with the naming for metrics. Right now we have `compute_parameters` , `profile_model_latency` and `weight_magnitude`. We should unify this such that they all follow the same pattern.

**Review by aaronkl — CHANGES_REQUESTED** [review]

File: whittle/metrics/latency.py (hunk 1 above)
Comment (rheasukthanker) [reply whittle/metrics/latency.py]: Addressed

File: whittle/metrics/latency.py (hunk 2 above)
Comment (rheasukthanker) [reply whittle/metrics/latency.py]: Addressed

File: whittle/metrics/latency.py (hunk 3 above)
Comment (rheasukthanker) [reply whittle/metrics/latency.py]: Addressed

File: whittle/metrics/latency.py (hunk 4 above)
Comment (rheasukthanker) [reply whittle/metrics/latency.py:28]: I don't think there should be an else, the cpu time is already computed before the loop, the if is only for legacy events, where some times could be saved on cuda.

File: whittle/metrics/latency.py (hunk 2 above)
Comment (rheasukthanker) [reply whittle/metrics/latency.py]: Will do that in a separate PR

File: whittle/metrics/latency.py (hunk 4 above)
Comment (aaronkl) [reply whittle/metrics/latency.py:28]: ah right, makes sense

**Review by aaronkl — APPROVED** [review]
LGTM

File: whittle/metrics/latency.py (hunk 2 above)
Comment (aaronkl) [reply whittle/metrics/latency.py]: added an issue #142

**Pull request merged**